### ISSUE TITLE
O2O integration branch based on CMSSW_8_0_7

### ISSUE THREAD

**Issue by mulhearn** [cms-l1t-offline/cmssw#298]
An integration branch for O2O developments has been created based on CMSSW_8_0_7 and l1t-integration-v47.0.  At this point these branches will diverge.  New features required for O2O will go into this branch.  New features may not change the results of the emulation:  we define O2O as working when it is able to bitwise reproduce l1t-integration-v47.0 with the new features turned on. 

Developers should base there O2O work on this branch and make PRs to this branch:

https://github.com/cms-l1t-offline/cmssw/tree/l1t-o2o-integration-CMSSW_8_0_7

CMSSW_8_0_7 is now available for o2o on the online system.  @kkotov is currently merging his updates for online O2O onto l1t-o2o-integration-v1.0.  Once ready, his changes will be added to this branch as well.

As of l1t-o2o-integration-v1.0:
- Latest common tools for XML parsing are included
- GMT ES producer is using online XML config files by default.
- l1t integration tests pass
- bitwise reproduces l1t-integration-v47.0

At l1t-o2o-integration-v2.0:
- O2O related update from uGT added
- Manual update of trigSystem to match @eparadas 

At l1t-o2o-integration-v3.0
- Added new CondFormats for EMTF, and helper classes.  Replace klunky pt assignment LUT emulation in helper.

v4.0 - stage1/stage2 calo params disentangled.

The recipe is:

```
cmsrel CMSSW_8_0_7
cd CMSSW_8_0_7/src
cmsenv
git cms-init
git remote add cms-l1t-offline git@github.com:cms-l1t-offline/cmssw.git
git fetch cms-l1t-offline
git cms-merge-topic --unsafe cms-l1t-offline:l1t-o2o-integration-v1.0
git cms-addpkg L1Trigger/L1TCommon
scram b -j 8
```

Until a fix is made for the XML files expected in local directory, you'll need to also do:

```
cp L1Trigger/L1TMuon/data/o2o/ugmt/*.xml .
```

The instructions from previous issue are:

Under L1TCommon, there are 4 new objects:
- _setting_ -> a class which takes care of each setting parameter (xml tag: param)
- _mask_ -> a class which takes care of each mask information (xml tag: mask, disabled)
- _xmlConfigReader_ -> a class which takes care of the XML file parsing
- _trigSystem_ -> a class which keeps the settings and the masks for the whole system

In order someone to create a trigSystem object properly, needs to do the follow:
1. Define the HW. By calling the method _addProcRole_, one should give as arguments the role and the processor: `void addProcRole(const std::string& role, const std::string& processor)`. This should be called as many times as the processors your system has.
2. Parse the XML files. By calling the method _configureSystem_, one should give as arguments the L1 key and the subsystem name: `void configureSystem(const std::string& l1HltKey, const std::string& subSysName)`. This method will be able to connect to the OMDS and retrieve the XML files. Since this is under development, we ask you to add the XML files manually. To do this, copy all the XML files, under CMSSW_$VERSION/src
   Then, you will need to edit the `L1Trigger/L1TCommon/src/trigSystem.cpp` lines 33 to 35, by adding the name of the top XML file and the key:
   - `_xmlRdr.readDOMFromFile("ugmt_top_config_p5.xml");`
   - `_xmlRdr.buildGlobalDoc("TestKey1");`
   - `_xmlRdr.readContexts("TestKey1", _sysId, *this);`
3. Once the above steps are done, you are ready to call the getters and retrieve the parameters you need. 
   Get/Check methods:
   1. `std::map<std::string, setting> getSettings (const std::string& processor)`
   2. `std::map<std::string, mask> getMasks(const std::string& proccessor)` 
   3. `bool isMasked(const std::string& proccessor, const std::string& id)` 
   4. `bool isProcEnabled(const std::string& proccessor)`
4. Explanation of the above methods:
   1. Returns a map with all the settings of a specific processor. The first element of the map is the id of the setting and the second is a setting object. With the method getValue, of the setting object, one can get the value in the type he prefers. 
      
      **Example**: 
         XML contains: `<param id="caloInputsDisable" type="bool">true</param>`
         Then in your ESProducer you can retrieve the value of the above, by doing:
         `std::map<std::string, l1t::setting> settings = trgSys.getSettings(aProcessor);`
         `disableCaloInputs = settings["disableCaloInputs"].getValue<bool>();`
      
         Some other possible returning methods are:
                                                                                  - `getVector()` which returns a Vector
                                                                                  - `getTableRows()` which returns a Table (xdata::Table in the XML file)
                                                                                  - `getLUT()` which returns a l1t::LUT object
         For these methods please have a look in the `setting.h` in order to call them properly. The getLUT has a few arguments. In order to retrieve a table setting here is an example:
      XML Contains:
      `<param id="regTable" type="table" delimiter="|">`
      `<columns>register_path| register_value</columns>`
                       `<types>string| uint</types>`
                           `<rows>`
                           `<row>payload.run.run_param.open_lut | 0</row>`
                           `</rows>`
                           `</param>`
      Then in your ESProducer you can retrieve the value of the above, by doing:
      `std::vector<l1t::tableRow>  tRow = settings["regTable"].getTableRows();`
      `for(auto it=tRow.begin(); it!=tRow.end(); it++)`
          `{`
            `if (it->getRowValue<std::string>("register_path").find("open_lut") != std::string::npos)`
             `std::cout << "Value is: " << it->getRowValue<bool>("register_value") << std::endl;`
            `}`
   2. Returns a map with all the masked links of a specific processor. The first element of the map is the id of the masked link and the second is a mask object. One can iterate on all the elements of the map and assign the masking to the emulator.
      **Example**: 
         XML contains: `mask id="inputPorts.Rx02"`
         Then in your ESProducer you can retrieve the value of the above, by doing:
         `std::map<std::string, l1t::setting> masks = trgSys.getMasks(aProcessor);`
         `for (auto& it: masks)`
         `std::cout << "Masked port: " << it.second.getId() << std::endl;`
   3. Returns true if the given id corresponds to a masked id. Keep in mind that the id should be as it is in the XML file, i.e. "inputPorts.Rx02".
   4. Returns true if the given processor is not disabled. This will be filled if only the "disable" tag is shown in the XML.

For a better understanding of all the above, you can have a look in the ESProducer of the uGMT `L1Trigger/L1TMuon/plugins/L1TMuonGlobalParamsESProducer.cc`

**Comment by puigh** [comment]
@mulhearn I don't see any of the recent uGT code in here, e.g. the PrescalesVetosHelper, L1TGlobalPrescalesVetosESProducer, etc. From where did you take the uGT?

**Comment by mulhearn** [comment]
Instructions updated above... you need also to do:
`cp L1Trigger/L1TMuon/data/o2o/ugmt/*.xml .`
for the uGMT example to work for now.

**Comment by mulhearn** [comment]
@puigh your code is in v2.0

**Comment by mulhearn** [comment]
@abrinke1 and @mrcarver as of v3.0, there are new CondFormats for the DT Forests used to emulate/derive the pt assignment.  Please have a look.  Its a major overhaul.  I demonstrated with several benchmarks that it bitwise reproduces the original version, but I left the original code in place in the pt assignment for now.  This code is going to need continued overhauls of this type, but for this branch I believe we can stop with the CondFormats.  This should be enough to provide the O2O of this payload, and we can handle the much needed cleanup of remaining end cap code outside the critical path.  Sound good?

**Comment by koskot77** [comment]
@mulhearn  v1.0 is merged with the L1T O2O and currently running at p5.
I only had to replace the default caloStage2Params_cfi for caloStage2Params_2016_v2_1_cfi.
If this is fine, I'll push my changes to v4.0

**Comment by mulhearn** [comment]
As mentioned in summary above, l1t-o2o-integration-v5.0 now includes all of the O2O-related PRs so far, including updates to uGT for masks and prescales, and the merging of the online O2O .

@koskot77:  you should bring your online code up to date with v5.0 now.

**Comment by kkotov** [comment]
In the light of the recent xml format change (https://github.com/cms-l1t-offline/cmssw/pull/320/) that cause the L1T O2O crash, I had to synchronize the online code with Thomas' private branch (thomreis_o2o-dev_CMSSW_8_0_7) that is, apart to the new parser, is no different from v5.0

**Comment by mulhearn** [comment]
PR #337 finally merged at tag l1t-o2o-integration-v6.0

**Comment by mulhearn** [comment]
This 809 branch is now deprecated.  Please move development to #341.
